File: 贪心算法/easy/1005.k-次取反后最大化的数组和.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=1005 lang=cpp
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
class Solution
{
public:
    // 排序函数，使用绝对值就行排序
    static bool cmp(int a, int b)
    {
        return abs(a) > abs(b);
    }
    int largestSumAfterKNegations(vector<int> &nums, int k)
    {
        int sum = 0;
        int size = nums.size();
        // 先通过绝对值排序
        sort(nums.begin(), nums.end(), cmp);

        for (int i = 0; i < size; i++)
        {
            // 优先将加大的负数转为正，同时消费k的数量：k--
            if (nums[i] < 0 && k != 0)
            {
                k--;
                nums[i] = abs(nums[i]);
            }

            sum += nums[i];
        }

        // 如果k依然存在，k是偶数，结果不变
        // k是奇数
        if (k % 2 != 0)
        {
            // 需要先减掉最后一个数
            sum -= nums[size - 1];
            // 然后将最后一个数的反向值
            sum -= nums[size - 1];
        }
        return sum;
    }
};
// @lc code=end

File: 贪心算法/easy/1005.k-次取反后最大化的数组和.ts
/*
 * @lc app=leetcode.cn id=1005 lang=typescript
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
function largestSumAfterKNegations(nums: number[], k: number): number {
  let sum = 0

  // 首先按照绝对值降序排序树组
  nums.sort((a, b) => Math.abs(b) - Math.abs(a))

  for (let i = 0; i < nums.length; i++) {
    // k的次数优先将绝对值较大的负数转化为证书
    if (nums[i] < 0 && k > 0) {
      nums[i] = - nums[i]
      k--
    }
  }

  // 剩下的次数是奇数，name需要将绝对值最小的数进行负数话
  if (k % 2 !== 0) {
    let lastIndex = nums.length - 1
    nums[lastIndex] = -nums[lastIndex]
  }

  // 累计求和
  for (let i = 0; i < nums.length; i++) {
    sum += nums[i]
  }

  return sum
};
// @lc code=end

